homeassistant/components/rituals_perfume_genie/sensor.py

```diff
@@ -0,0 +1,147 @@
+    diffusers = hass.data[DOMAIN][config_entry.entry_id][DEVICES]
+    coordinators = hass.data[DOMAIN][config_entry.entry_id][COORDINATORS]
+    entities = []
+    for hublot, diffuser in diffusers.items():
+        coordinator = coordinators[hublot]
+        entities.append(DiffuserPerfumeSensor(diffuser, coordinator))
+        entities.append(DiffuserFillSensor(diffuser, coordinator))
+        entities.append(DiffuserWifiSensor(diffuser, coordinator))
+        if BATTERY in diffuser.data[HUB][SENSORS]:
+            entities.append(DiffuserBatterySensor(diffuser, coordinator))
+
+    async_add_entities(entities)
+
+
+class DiffuserPerfumeSensor(DiffuserEntity):
+    """Representation of a diffuser perfume sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the perfume sensor."""
+        super().__init__(diffuser, coordinator, PERFUME_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the perfume sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][PERFUME][ID] == PERFUME_NO_CARTRIDGE_ID:
+            return "mdi:tag-remove"
+        return "mdi:tag-text"
+
+    @property
+    def state(self):
+        """Return the state of the perfume sensor."""
+        return self.coordinator.data[HUB][SENSORS][PERFUME][TITLE]
+
+
+class DiffuserFillSensor(DiffuserEntity):
+    """Representation of a diffuser fill sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the fill sensor."""
+        super().__init__(diffuser, coordinator, FILL_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the fill sensor icon."""
+        if self.coordinator.data[HUB][SENSORS][FILL][ID] == FILL_NO_CARTRIDGE_ID:
+            return "mdi:beaker-question"
+        return "mdi:beaker"
+
+    @property
+    def state(self):
+        """Return the state of the fill sensor."""
+        return self.coordinator.data[HUB][SENSORS][FILL][TITLE]
+
+
+class DiffuserBatterySensor(DiffuserEntity):
+    """Representation of a diffuser battery sensor."""
+
+    def __init__(self, diffuser, coordinator):
+        """Initialize the battery sensor."""
+        super().__init__(diffuser, coordinator, BATTERY_SUFFIX)
+
+    @property
+    def icon(self):
+        """Return the battery sensor icon."""
+        return {
+            "battery-charge.png": "mdi:battery-charging",
+            "battery-full.png": "mdi:battery",
+            "battery-75.png": "mdi:battery-50",
+            "battery-50.png": "mdi:battery-20",
+            "battery-low.png": "mdi:battery-alert",
+        }[self.coordinator.data[HUB][SENSORS][BATTERY][ICON]]
+
+    @property
+    def state(self):
+        """Return the state of the battery sensor."""
+        return self.coordinator.data[HUB][SENSORS][BATTERY][TITLE]
+
+    @property
+    def device_class(self):
+        """Return the class of the battery sensor."""
+        return DEVICE_CLASS_BATTERY
+
+    @property
+    def device_state_attributes(self):
```

Comment:
These have been renamed to `extra_state_attributes`

Reply:
```suggestion
    def extra_state_attributes(self):
```